cli/flox/src/commands/gc.rs

```diff
@@ -1,23 +1,291 @@
+            while let Ok(new_progress) = receiver.recv() {
+                if let GcProgress::Done(ref msg) = new_progress {
+                    done_msg = msg.clone();
+                }
+                // This dance is to make sure that the span (and hence the spinner)
+                // stays open until the _start of the next loop body_,
+                // rather than closing at the end of the current loop body
+                // (which is immediately, basically).
+                if let Some(guard) = inner_span.take() {
+                    drop(guard);
+                }
+                inner_span = Some(
+                    info_span!(
+                        parent: &parent_span,
+                        "gc_progress",
+                        progress = new_progress.msg().unwrap_or("Finishing".to_string())
+                    )
+                    .entered(),
+                );
+                debug!(
+                    state = new_progress.state_name(),
+                    "updated spinner with new state"
+                );
+            }
+            debug!("gc log sender hung up, exiting spinner thread");
+
+            // Print a message indicating how many store paths were deleted
+            // and how much disk space was reclaimed.
+            if !done_msg.is_empty() {
+                message::info(done_msg);
+            }
+
+            Ok(())
+        });
+        debug!("spawned spinner update thread");
+
+        // This thread collects and parses logs from the output of the `nix store gc`
+        // command. It keeps track of the GC state so that we can inform the
+        // spinner thread only when we need to.
+        let reader_thread: std::thread::ScopedJoinHandle<
+            '_,
+            std::result::Result<(), anyhow::Error>,
+        > = s.spawn(move || {
+            for line in reader.lines() {
+                let line = line.context("Failed to read GC process output")?;
+                if let Some(new_progress) = gc_progress.new_progress_state(&line) {
+                    debug!(line, "got new progress state");
+                    sender
+                        .send(new_progress.clone())
+                        .context("Background thread exited early")?;
+                    gc_progress = new_progress;
+                }
+            }
+            debug!("consumed all output, exiting gc reader thread");
+            Ok(())
+        });
+        debug!("spawned gc reader thread");
+
+        // Wait for the threads to exit.
+        spinner_thread
+            .join()
+            .map_err(|_| anyhow!("Background thread panicked"))??;
+        debug!("joined spinner update thread");
+        reader_thread
+            .join()
+            .map_err(|_| anyhow!("Background thread panicked"))??;
+        debug!("joined gc reader thread");
+        Ok::<(), anyhow::Error>(())
+    })?;
+    Ok(())
+}
+
+// The order of log messages during GC
```

Comment:
These will get deleted, they were just notes to myself as I was reading the debug output of the `nix store gc` command.